src/compiler/crystal/semantic/call_error.cr

```diff
@@ -194,6 +176,311 @@ class Crystal::Call
+    all_names = call_errors.flat_map(&.names).uniq
+    all_names_in_all_overloads = all_names.select do |name|
+      call_errors.all? &.names.includes?(name)
+    end
+    return if all_names_in_all_overloads.empty?
+
+    error_message = String.build do |str|
+      quoted_names = all_names.map { |name| "'#{name}'" }
+
+      if quoted_names.size == 1
+        str << "argument for parameter #{quoted_names.first} already specified"
+      else
+        str << "arguments for parameters #{quoted_names.join ", "} already specified"
+      end
+
+      str.puts
+      str.puts
+      str << "Overloads are:"
+      append_matches(defs, arg_types, str)
+    end
+
+    raise error_message, inner: inner_exception
+  end
+
+  private def check_wrong_number_of_arguments(call_errors, owner, defs, def_name, arg_types, named_args_types, inner_exception)
+    return unless call_errors.all?(WrongNumberOfArguments)
+
+    raise_matches_not_found_named_args(owner, def_name, defs, arg_types, named_args_types, inner_exception)
+  end
+
+  private def check_arguments_type_mismatch(call_errors, owner, defs, def_name, arg_types, named_args_types, inner_exception)
+    call_errors = call_errors.select(ArgumentsTypeMismatch)
+    return if call_errors.empty?
+
+    call_errors = call_errors.map &.as(ArgumentsTypeMismatch)
+    argument_type_mismatches = call_errors.flat_map(&.errors)
+
+    all_indexes_or_names = argument_type_mismatches.map(&.index_or_name).uniq
+    indexes_or_names_in_all_overloads = all_indexes_or_names.select do |index_or_name|
+      call_errors.all? &.errors.any? &.index_or_name.==(index_or_name)
+    end
+
+    return if indexes_or_names_in_all_overloads.empty?
+
+    # Only show the first error. We'll still list all overloads.
+    index_or_name = indexes_or_names_in_all_overloads.first
+
+    mismatches = argument_type_mismatches.select(&.index_or_name.==(index_or_name))
+    expected_types = mismatches.map(&.expected_type).uniq.sort_by(&.to_s)
+    actual_type = mismatches.first.actual_type
+
+    arg =
+      case index_or_name
+      in Int32
+        args[index_or_name]?
+      in String
+        named_args.try &.find(&.name.==(index_or_name))
+      end
+
+    error_message = String.build do |str|
+      argument_description =
+        case index_or_name
+        in Int32
+          "##{index_or_name + 1}"
+        in String
+          "'#{index_or_name}'"
+        end
+
+      str << "expected argument "
+      str << argument_description
+      str << " to '"
+      str << full_name(owner, def_name)
+      str << "' to be "
+      if expected_types.size == 1
+        str << expected_types.first
+      else
+        expected_types.each_with_index do |expected_type, i|
+          if i == expected_types.size - 1
+            str << " or "
+          elsif i > 0
+            str << ", "
+          end
+          str << expected_type
+        end
+      end
+      str << ", not "
+      str << actual_type
```

Comment:
This section seems a maybe bit strongly optimized for avoiding intermediary allocations from string interpolations? As a result it's a bit hard to read when the code for constructing a sentence is spread over 20 lines.
I think the part for writing the array in sentence form with an "or" at the end should definitely be moved out. This is a very common use case and can easily be needed elsewhere.
This might even be something for stdlib, similar to Rail's core extension [`Array#to_sentence`](https://api.rubyonrails.org/classes/Array.html#method-i-to_sentence)?

Reply:
I'll move it to a helper method in that file. I think this is the only place where it's needed so far. Or where you thinking about adding this to Array as a core_ext thing?

Reply:
Just a helper method would be fine.

But we could think about adding an implementation for `Array` in stdlib.

Reply:
Done!
Yeah, introducing it for Array might be good. I have mixed feelings, though: it feels like front-end-specific logic, though I can imagine CLIs and many applications using it.